src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ThisEscapeAnalyzer.java

```diff
@@ -0,0 +1,1475 @@
+        DiagnosticPosition[] previous = null;
+        for (DiagnosticPosition[] warning : this.warningList) {
+
+            // Skip duplicates
+            if (previous != null && extendsAsPrefix.test(previous, warning))
+                continue;
+            previous = warning;
+
+            // Emit warnings showing the entire stack trace
+            JCDiagnostic.Warning key = Warnings.PossibleThisEscape;
+            int remain = warning.length;
+            do {
+                final DiagnosticPosition pos = warning[--remain];
+                this.log.warning(Lint.LintCategory.THIS_ESCAPE, pos, key);
+                key = Warnings.PossibleThisEscapeLocation;
+            } while (remain > 0);
+        }
+        this.warningList.clear();
+    }
+
+    private void analyzeStatements(List<JCStatement> stats) {
+        for (JCStatement stat : stats) {
+
+            // Analyze statement
+            this.scan(stat);
+
+            // Capture any pending warning generated
+            if (this.copyPendingWarning())
+                break;                      // report at most one warning per constructor
+        }
+    }
+
+    @Override
+    public void scan(JCTree tree) {
+
+        // Check node
+        if (tree == null || tree.type == Type.stuckType)
+            return;
+
+        // Sanity check
+        Assert.check(this.checkInvariants(true, false));
+
+        // Can this expression node possibly leave a 'this' reference on the stack?
+        final boolean referenceExpressionNode;
+        switch (tree.getTag()) {
+        case CASE:
+        case SWITCH_EXPRESSION:
+        case CONDEXPR:
+        case YIELD:
+        case APPLY:
+        case NEWCLASS:
+        case NEWARRAY:
+        case LAMBDA:
+        case PARENS:
+        case ASSIGN:
+        case TYPECAST:
+        case INDEXED:
+        case SELECT:
+        case REFERENCE:
+        case IDENT:
+        case NULLCHK:
+        case LETEXPR:
+            referenceExpressionNode = true;
+            break;
+        default:
+            referenceExpressionNode = false;
+            break;
+        }
+
+        // Scan node
+        super.scan(tree);
+
+        // Sanity check
+        Assert.check(this.checkInvariants(true, referenceExpressionNode));
+
+        // Discard any direct 'this' reference that's incompatible with the target type
+        if (referenceExpressionNode) {
+
+            // We treat instance methods as having a "value" equal to their instance
+            Type type = tree.type;
+            final Symbol sym = TreeInfo.symbolFor(tree);
+            if (sym != null &&
+                sym.kind == MTH &&
+                (sym.flags() & Flags.STATIC) == 0) {
+                type = sym.owner.type;
+            }
+
+            // If the expression type is incompatible with 'this', discard it
+            if (type != null && !this.isSubtype(this.targetClass.sym.type, type))
```

Comment:
No because (for example) what if you cast?

The thing you're casting might be compatible, but after the cast it might become incompatible.

Reply:
Uhm. Turns out I probably did not understand the filter correctly, and now I'm more dubious about what it actually does. Say you have this hierarchy:

```
interface A { }
class B {
     B() {
          A a = (A)this;
          ...
      }
 }
 class C extends B implements A { }
 ```

Pathological case, I know. But the filtering will end up dropping the expression Ref on the floor, right? (because B and A are unrelated).

Reply:
> But the filtering will end up dropping the expression Ref on the floor, right? (because B and A are unrelated).

Ah, I see what you mean.

Here's a more complete example:
```java
public class CastLeak {

    public CastLeak() {
        ((CastLeak)(Runnable)this).mightLeak();
    }

    public void mightLeak() {
    }
}
```
That would be a leak for any subclass that implements `Runnable`. Yet no warning is generated.

So the filtering by expression type is going to potentially create false negatives. But it also eliminates a bunch of false positives. And the false negatives are probably all somewhat pathological like the example above.

So I still think it's worth doing.

Reply:
>  Ok - I thought false negative was the thing to absolutely avoid - and that was the no. 1 concern.

You're right. I think at the time I reasoned that it would be unusual enough for the type of an expression to start as an instanceof X, then change to not being an instanceof X, and then change back, that it was worth it to go ahead and filter these out. But admittedly that was based on intuition, not science.

> I think a possible approach to keep both the filtering and the code more or less similar to what you have, is to save the type of the expression in the ExprRef. Then, I believe that, at the end of scan() you can just get whatever type is there, and check that it's the correct one, if not drop it.

That's a nice idea... thanks. I'll play around with it some.

Reply:
I was curious how much of a difference this type filtering makes, so I built the JDK with and without it.

The results were identical except for one case:
```java
package java.lang.invoke;
...
public abstract sealed class CallSite permits ConstantCallSite, MutableCallSite, VolatileCallSite {
...
    CallSite(MethodType targetType, MethodHandle createTargetHook) throws Throwable {
        this(targetType); // need to initialize target to make CallSite.type() work in createTargetHook
        ConstantCallSite selfCCS = (ConstantCallSite) this;
        MethodHandle boundTarget = (MethodHandle) createTargetHook.invokeWithArguments(selfCCS);
        setTargetNormal(boundTarget); // ConstantCallSite doesn't publish CallSite.target
        UNSAFE.storeStoreFence(); // barrier between target and isFrozen updates
    }
```
When we do type filtering, `(ConstantCallSite) this` causes the 'this' reference to be discarded so no leak is reported on the next line for `invokeWithArguments(selfCCS)`. Just a side note, there is also a leak on the next line because final method `setTargetNormal()` invokes `MethodHandleNatives.setCallSiteTargetNormal(this, newTarget)`, so a leak does get reported anyway even with type filtering.

When type filtering is disabled, we report a leak at `invokeWithArguments(selfCCS)` - which is accurate.

So what did we learn?

First it looks like type filtering has very minimal effect. I think this is because it requires some version of two casts in a row in and out of type compatibility, and this is probably very rare.

But looking at the one data point we do have, the type filtering did in fact cause a false negative.

And when building the entire JDK, it causes zero net new false positives.

So to me this is evidence that we should just remove the type filtering altogether...

@vicente-romero-oracle your thoughts?

Reply:
I believe the conclusion you reached is sound - it doesn't look like type filtering adds much, and it can be actively harmful.